ls1intum/Artemis
# `Development`: Add more features toggles to the programming language service

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises

### Motivation and Context
This PR adds three new features to the `ProgrammingLanguageFeature` class. These include:
- `testwiseCoverageAnalysisSupported`: Whether the CI system can create coverage reports for a programming language. This is only allowed for Java and Kotlin exercises and until now was tested without using the `ProgrammingLanguageFeatureService`.
- `auxiliaryRepositoriesSupported`: Whether auxiliary repositories can be created for a programming language/CI system combination. (currently `true` for all CI systems and programming languages, see explanation below)
- `publishBuildPlanUrlAllowed`: Whether the URL linking to the build plan on Bamboo/Jenkins/GitLab may be published to the student. (currently `true` for all CI systems and programming languages, see explanation below)

`auxiliaryRepositoriesSupported` and `publishBuildPlanUrlAllowed` are currently `true` for all CI systems and programming languages and would thus not require being added to the `ProgrammingLanguageFeature` system. However, they will be needed for the upcoming local CI system. Auxiliary repositories will not be supported initially and publishing a build plan URL does not make sense for a CI system without a graphical user interface. This PR thus helps in shrinking the size of the big local CI PR.

### Description
In addition to adding three new features, the `ProgrammingLanguageFeature` class was changed to a record.

### Steps for Testing
Prerequisites:
- 1 Instructor

1. Log in to Artemis
2. Navigate to Course Management -> Exercises for one course
3. Click "Generate new Programming Exercise"
4. Make sure the different features are switched on and off correctly depending on the programming language according to the table here: https://docs.artemis.cit.tum.de/user/exercises/programming/#features (e.g. you can select sequential test runs for Java but not for Swift, be aware that the table is not entirely correct if you deploy on TS2 (using Jenkins as its CI system)).

### Review Progress

#### Code Review
#### Manual Tests

### Test Coverage
| Class/File | Branch | Line |
|------------|-------:|-----:|
| ProgrammingExerciseResource.java | 79% | 89% |
| programming-exercise-update.component.ts | 72.18% | 79.12% |
| programming-language-feature.service.ts | 75% | 90.9% |
